src/main/java/org/motechproject/prevac/service/impl/PrimeVaccinationScheduleServiceImpl.java

```diff
@@ -55,47 +55,78 @@
+
+        Clinic clinic = primeVisit.getClinic();
 
         if (clinic != null && !ignoreLimitation) {
             checkNumberOfPatients(dto, clinic);
         }
 
-        return new PrimeVaccinationScheduleDto(updateVisitWithDto(primeDetails, screeningDetails, dto));
+        return new PrimeVaccinationScheduleDto(primeVisit);
     }
 
     @Override
     public List<PrimeVaccinationScheduleDto> getPrimeVaccinationScheduleRecords() {
         List<PrimeVaccinationScheduleDto> primeVacDtos = new ArrayList<>();
-        createEmptyVisitsForSubjectWithoutPrimeVacDate();
-        List<Visit> detailsList = visitBookingDetailsDataService
-                .findByParticipantNamePrimeVaccinationDateAndVisitTypeAndBookingPlannedDateEq(".", null, VisitType.PRIME_VACCINATION_DAY, null);
+        // Get all subjects
+        List<Subject> subjects = subjectDataService.findByPrimerVaccinationDate(null);
 
-        for (Visit details : detailsList) {
-            primeVacDtos.add(new PrimeVaccinationScheduleDto(details));
+        for (Subject subject : subjects) {
+            if (subject.getVisits().isEmpty()) {
+                primeVacDtos.add(new PrimeVaccinationScheduleDto(subject));
+            }
         }
 
         return primeVacDtos;
     }
 
+    private Visit createScreeningVisitFromDto(PrimeVaccinationScheduleDto dto, Subject subject) {
+        Visit screeningVisit = new Visit();
+        screeningVisit.setType(VisitType.SCREENING);
+        screeningVisit.setSubject(subject);
+        screeningVisit.setDate(dto.getBookingScreeningActualDate());
+        return screeningVisit;
+    }
+
+    private Visit createPrimeVacVisitFromDto(PrimeVaccinationScheduleDto dto, Subject subject) {
+        Visit primeVisit = new Visit();
+        primeVisit.setSubject(subject);
+        primeVisit.setType(VisitType.PRIME_VACCINATION_DAY);
+        primeVisit.setStartTime(dto.getStartTime());
+        primeVisit.setEndTime(calculateEndTime(dto.getStartTime()));
+        primeVisit.setDateProjected(dto.getDate());
+        primeVisit.getSubject().setFemaleChildBearingAge(dto.getFemaleChildBearingAge());
+        primeVisit.setIgnoreDateLimitation(dto.getIgnoreDateLimitation());
+        return primeVisit;
+    }
+
     private Visit updateVisitWithDto(Visit primeDetails, Visit screeningDetails,
                                      PrimeVaccinationScheduleDto dto) {
         primeDetails.setStartTime(dto.getStartTime());
         primeDetails.setEndTime(calculateEndTime(dto.getStartTime()));
-        primeDetails.setBookingPlannedDate(dto.getDate());
+        primeDetails.setDateProjected(dto.getDate());
         primeDetails.getSubject().setFemaleChildBearingAge(dto.getFemaleChildBearingAge());
         primeDetails.setIgnoreDateLimitation(dto.getIgnoreDateLimitation());
 
-        screeningDetails.setBookingActualDate(dto.getBookingScreeningActualDate());
+        screeningDetails.setDate(dto.getBookingScreeningActualDate());
```

Comment:
bookingScreeningActualDate should not be used anymore, use actualScreeningDate instead